### PR TITLE
Show IP that connected to and IP that failed

### PR DESCRIPTION
**Pull request by dHannasch**
This is built atop https://github.com/ray-project/ray/pull/12022, which is built atop https://github.com/ray-project/ray/pull/12016, so this *looks* very large and complicated at the moment but there's really only one simple change happening here. It'll look simpler if those are dealt with first.

## Why are these changes needed?
This emits a warning when we receive a shard address from RedisShards that does not appear to be accessible from this node.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
BTW, @dHannasch one thing that would be really useful for reviewing is if you could post the "before" and "after" of each error message that you change.

**Comment by dHannasch** [comment]
> post the "before" and "after" of each error message that you change.

I'm not sure what you mean? This emits a warning when we receive a shard address from RedisShards that does not appear to be accessible from this node. Currently, there isn't one. There is no "before". I promise, that really truly is all this branch is doing.
```
      Status portReachable = shard_contexts_[i]->PingPort(addresses[i], ports[i]);
      if (!portReachable.ok()) {
        RAY_LOG(WARNING) << "Successfully connected to " << options_.server_ip_ << ":"
                         << options_.server_port_ << ", but not to the shard address it "
                         << "sent: " << portReachable.message();
      }
```

Like it says above, it would probably be easier to start with https://github.com/ray-project/ray/pull/12016. https://github.com/ray-project/ray/pull/12016 is much easier to read on its own, since it's only doing one thing (extracting a function). https://github.com/ray-project/ray/pull/12016 doesn't change *any* messages, just extracts that function for use.

**Comment by richardliaw** [comment]
Ah, I guess I didn't read this PR carefully (and assumed this was a modification of an existing warning message). Thanks for the detailed response! 

In any case, I guess for me, seeing the (new/modified) error message printed will help me see things in context (like rendering a demo). Does that make sense? It also just helps me validate that the code actually works as I expect without needing to build your branch myself :)

File: src/ray/gcs/redis_client.cc [hunk 1]
```diff
@@ -132,6 +132,12 @@ Status RedisClient::Connect(std::vector<boost::asio::io_service *> io_services)
       boost::asio::io_service &io_service = *io_services[io_service_index];
       // Populate shard_contexts.
       shard_contexts_.push_back(std::make_shared<RedisContext>(io_service));
+      Status portReachable = shard_contexts_[i]->PingPort(addresses[i], ports[i]);
+      if (!portReachable.ok()) {
+        RAY_LOG(WARNING) << "Successfully connected to " << options_.server_ip_ << ":"
+                         << options_.server_port_ << ", but not to the shard address it "
+                         << "sent: " << portReachable.message();
+      }
```
Comment (richardliaw) [inline src/ray/gcs/redis_client.cc:140]: thanks for this PR, @dHannasch -- I merged in 12026 for you.

Can you actually help provide an example for where this is useful?

**Review by richardliaw — COMMENTED** [review]
In this PR, can you add a test for PingPort?

**Comment by dHannasch** [comment]
> In this PR, can you add a test for PingPort?

Can do, sorry for the delay. Is there a place `Connect()` is tested? I'm not finding it. I assume the PingPort test should go right next to it and work the same way?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**